### PR TITLE
Shaked

### PR DESCRIPTION
**Pull request by shakederin**
DAY 1 mission A

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -1 +1,82 @@
-// server code
\ No newline at end of file
+// server code
+const express = require('express');
+const morgan = require('morgan')
+const app = express();
+const data = require("./database")
```
Comment (ElayGelbart) [inline src/server.js]: Put all your imports together

File: src/server.js [hunk 2]
```diff
@@ -1 +1,82 @@
...
+    return JSON.stringify(req.body);
+})
+app.use(morgan(":method :url :status :res[content-length] - :response-time ms :data", {
+    skip : function(req,res){
+       return req.method !== "POST" 
+    }    
+}))
+
+app.get("/api/persons",  function (req,res){
+    res.send(data.data);
```
Comment (ElayGelbart) [inline src/server.js]: Bad Name

File: src/server.js [hunk 3]
```diff
@@ -1 +1,82 @@
...
+});
+
+app.get("/info", function (req,res){
+    res.send({Contacts: data.data.length, date:new Date()})
```
Comment (ElayGelbart) [inline src/server.js]: Not right Format

File: src/server.js [hunk 4]
```diff
@@ -1 +1,82 @@
...
+});
+
+app.get("/api/persons/:id", function (req,res){
+    const id = req.params.id;
+    let contact;
+    for (let i =0; i<data.data.length; i++){
```
Comment (ElayGelbart) [inline src/server.js]: Could use for(value of array)

File: src/server.js [hunk 5]
```diff
@@ -1 +1,82 @@
...
+        if(+data.data[i].id === +id){
```
Comment (ElayGelbart) [inline src/server.js]: Not use `Math.random()`

File: src/server.js [hunk 6]
```diff
@@ -1 +1,82 @@
...
+            contact = data.data[i];
+            res.send(contact);
+            return
+        }
+    }
+    res.sendStatus(404);
```
Comment (ElayGelbart) [inline src/server.js]: what about special msg

File: src/server.js [hunk 7]
```diff
@@ -1 +1,82 @@
...
+})
+
+app.delete("/api/persons/:id", function (req,res){
+    const id = req.params.id;
+    for (let i =0; i<data.data.length; i++){
+        if(+data.data[i].id === +id){
+            console.log(data.data);
```
Comment (ElayGelbart) [inline src/server.js]: console log in publish code

File: src/server.js [hunk 8]
```diff
@@ -1 +1,82 @@
...
+        return;
+    }
+    let idNumber = 0;
+    for(let i = 0; i < data.data.length; i++){
+        if(data.data[i].name === newContactName){
+            res.status(403).send("That name is already exists in the phonebook");
+            return;
+        }
+        if(data.data[i].id > idNumber){
+            idNumber = data.data[i].id
```
Comment (ElayGelbart) [inline src/server.js]: ?????

**Review by ElayGelbart — CHANGES_REQUESTED** [review]
Fix

File: frontend/index.html [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css"
+        rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
+    <link rel="stylesheet" href="./style.css">
+    <title>Document</title>
+</head>
+<body>
+    <nav id="navBar">
```
Comment (ElayGelbart) [inline frontend/index.html:13]: why id ?

File: frontend/index.html [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            <div class="col-sm-7">
+              <input id="pnumber" type="number" class="form-control" placeholder="Phone Number" aria-label="City">
+            </div>
+            <button id="addBtn">ADD</button>
+            <div id="errormsg"></div>
+        </div>
+    </div>
+    
+    <script src="./index.js"></script>
+    <script src="https://unpkg.com/axios/dist/axios.min.js"></script>
```
Comment (ElayGelbart) [inline frontend/index.html:47]: axios in head

File: frontend/index.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+const errormsg = document.getElementById("errormsg");
+const book = document.getElementById("book");
+const showAll = document.getElementById("showAll");
+const hideAll = document.getElementById("hideAll");
+
+addBtn.addEventListener("click", async function(){
+    const firstName = fName.value;
+    const lastName = lName.value;
+    const phoneNumber = pNumber.value;
+    const fullName = firstName + " " + lastName;
```
Comment (ElayGelbart) [inline frontend/index.js:17]: `${}`

File: frontend/index.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    const tdNumber = document.createElement("td")
+    const favorite = document.createElement("td")
+    const deleteOption = document.createElement("td")
+    tdName.innerText = name;
+    tdNumber.innerText = number;
+    favorite.innerText = "Favorites";
+    deleteOption.innerText = "Delete";
+    favorite.classList.add("cursorPink");
+    deleteOption.classList.add("cursorRed");
+    favorite.addEventListener("click", function(){makeFavorite(contact)});
```
Comment (ElayGelbart) [inline frontend/index.js:73]: why Favourite

File: frontend/index.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    deleteOption.addEventListener("click", async function(){await deleteContact(contact, name)});
+    contact.classList.add("dropDownDiv");
+    tdName.append(favorite);
+    tdName.append(deleteOption);
+    contact.append(tdName);
+    contact.append(tdNumber);
+    book.append(contact);
+}
+
+async function deleteContact(contact, name){
```
Comment (ElayGelbart) [inline frontend/index.js:83]: what is contact? element? string? number?

File: frontend/index.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    contact.remove();
+    let id;
+    const data = await axios.get(`${myAPI}/api/persons`);
```
Comment (ElayGelbart) [inline frontend/index.js:86]: slow, why get and also delete

File: frontend/index.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    for(let i = 0; i < data.data.length ; i++){
+        if(data.data[i].name === name){
```
Comment (ElayGelbart) [inline frontend/index.js:88]: data data

File: frontend/style.css [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    grid-row: 4/10;
+    font-size: 150%;
+    overflow-y: auto; 
+    overflow-x: hidden; 
+    
+} 
+
+
+#errormsg{
+    margin-bottom: 5%;
```
Comment (ElayGelbart) [inline frontend/style.css:35]: why %

File: frontend/style.css [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    cursor: pointer;
+    background-color: red;
+}
+
+#hideAll{
+    cursor: pointer;
+    text-align: center;
+    visibility: hidden;
+    position: absolute;
+    left: 47%;
```
Comment (ElayGelbart) [inline frontend/style.css:76]: %?

Comment (ElayGelbart) [reply frontend/style.css:76]: `display:flex`

File: toDataBase/mongo.js [hunk 18]
```diff
@@ -0,0 +1,70 @@
+const mongoose = require('mongoose');
+require('dotenv').config();
```
Comment (ElayGelbart) [inline toDataBase/mongo.js]: if you use argv you dont need dotenv

File: webpack.config.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+// const HtmlWebpackPlugin = require('html-webpack-plugin');
+// const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+
+module.exports = {
+  entry: './frontend/index.js',
+  output: {
+    filename: 'bundle.js',
+    path: path.resolve(__dirname, 'dist'),
+  }
+//   devtool: 'eval-source-map',
```
Comment (ElayGelbart) [inline webpack.config.js:11]: delete `//`

**Review by ElayGelbart — COMMENTED** [review]
enjoy

File: .env [hunk 20]
```diff
@@ -0,0 +1,5 @@
+DATABASE = "mongodb+srv://shakederin:shaked12345675@cluster0.7gngs.mongodb.net/phonebook?retryWrites=true&w=majority"
```
Comment (ehood-goren) [inline .env:1]: Dont commit .env file

File: models/note.js [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+.then(()=>{console.log("DB connected")})
+.catch((error)=>{'error connecting to MongoDB:', error.message});
+
+const phoneBookSchema = new mongoose.Schema({
+    id : {
+      type :  Number,
+      unique : true,
+      required: true
+    },
+    name: String,
```
Comment (ehood-goren) [inline models/note.js:22]: require?

File: models/note.js [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+
+// const key = process.env.DATABASE1;
+// const key1 = process.env.DATABASE2;
+// const password = process.argv[2]
+// const connectionString = key+password+key1;
+// program(process.argv[2], process.argv[3], process.argv[4])
+
+const connectionString = process.env.DATABASE;
+
+mongoose.connect(connectionString)
```
Comment (ehood-goren) [inline models/note.js:12]: Don't connect in model

File: models/note.js [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+.then(()=>{console.log("DB connected")})
+.catch((error)=>{'error connecting to MongoDB:', error.message});
+
+const phoneBookSchema = new mongoose.Schema({
```
Comment (ehood-goren) [inline models/note.js:16]: separate model

File: models/note.js [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+    id : {
+      type :  Number,
+      unique : true,
+      required: true
+    },
+    name: String,
+    phoneNumber: String
+})
+
+function program(password, name, phoneNum){
```
Comment (ehood-goren) [inline models/note.js:26]: this doesn't belong here

File: src/server.js [hunk 25]
```diff
@@ -1 +1,103 @@
-// server code
\ No newline at end of file
+const express = require('express');
+const path = require('path')
+const Note = require('../models/note');
```
Comment (ehood-goren) [inline src/server.js:3]: Note?

File: src/server.js [hunk 26]
```diff
@@ -1 +1,103 @@
...
+const cors = require("cors")
+require('dotenv').config();
+const app = express();
+const port = process.env.PORT || 3001;
+const connectionString = process.env.DATABASE;
```
Comment (ehood-goren) [inline src/server.js:8]: Not used

File: src/server.js [hunk 27]
```diff
@@ -1 +1,103 @@
...
+    const id = req.params.id;
+    try {
+        const contact = await Note.find({id: id});
+        if(contact[0]){
+            res.send(contact[0]);
+        }else{
+            throw "Wrong Id";
+        }
+    } catch (error) {
+        res.status(400).send("Cant find person");
```
Comment (ehood-goren) [inline src/server.js:43]: Use error handler

File: src/server.js [hunk 28]
```diff
@@ -1 +1,103 @@
...
+            throw "error";
+        }
+       
+    } catch (err) {
+        res.status(400).send("Cant find person");  
+    }       
+ })
+
+app.post("/api/persons", async function (req,res){
+    if (req.body === undefined) {
```
Comment (ehood-goren) [inline src/server.js:64]: !req.body

File: src/server.js [hunk 29]
```diff
@@ -1 +1,103 @@
...
+        return res.status(400).json({ error: 'content missing' })
```
Comment (ehood-goren) [inline src/server.js:65]: Use the same format for errors

File: src/server.js [hunk 30]
```diff
@@ -1 +1,103 @@
...
+    const newContact = new Note({
+        "id": newId,
+        "name": newContactName,
+        "phoneNumber": newContactNumber
+    })
+    newContact.save().then((savedContact)=>{
+        res.send(`${savedContact.name} successfully add to your phonebook!`);
+    })
+    } catch (error) {
+        res.send(error);
```
Comment (ehood-goren) [inline src/server.js:91]: Another error format

**Review by ehood-goren — CHANGES_REQUESTED** [review]

**Pull request closed without merging**